### ISSUE TITLE
A few comments on the RNN implementation

### ISSUE THREAD

**Issue by coltekin** [cicl2018/HateEvalTeam#1]
As promised yesterday, I looked at your RNN code. I have a few comments below, but for reference, here is your model definition:

```
    recurrent_model = Sequential()
    recurrent_model.add(LSTM(units=64, return_sequences=True))
    recurrent_model.add(LSTM(units=64, return_sequences=True))
    recurrent_model.add(Dropout(0.2))
    recurrent_model.add(Activation('tanh'))
    recurrent_model.add(Flatten())
    recurrent_model.add(Dense(units=1, activation='sigmoid'))
    recurrent_model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    recurrent_model.fit(train_x, train_y, epochs=30)
```

- I suggest using an embedding layer. Even if you are using characters, it will reduce the model size, and there may be features of characters that may put them into a reasonable space where similarities matter (e.g., your model may potentially learn the similarity/equivalence of uppercase and lowercase characters). If you do so, also make sure to 'mask_zero', so that your model ignores the padding.
- You are returning all the sequences from LSTMs, so your LSTMs output (longest_sentence x 64) vectors for each training/testing instance. This is fine when you are stacking them, but probably you do not want to do this (unless you want to implement some attention mechanism) on the final LSTM. Just using the 64-d final activation should be fine.
- `recurrent_model.add(Activation('tanh'))` is probably not doing something you intended. The LSTM cells include all the activation functions they need. Here, only thing you do is you pass the output of the LSTM from one more `tanh` activation. This will slightly compress the output of LSTM (which is itself passed thorough `tanh`), but I am not sure what benefit this could have. Note that this is not a "layer" in the usual sense, there are no weights associated with it.
- You may also consider applying dropout to LSTM inputs/recurrent connections. I heard conflicting reports on benefit of dropout on recurrent connections, but applying dropout to inputs is a standard procedure. Normally, you can do this by passing arguments to the layer. Dropout after the LSTM may also be useful, but I think you can get better overfitting  prevention by applying it "in" the LSTM cells.
- In the long run, I suggest using a 'early stopping' callback and trying to optimize the network on F-score to match the shared task evaluation. You'd need to write your own callback for this, but it is doable with Keras too.

**Comment by guzimanis** [comment]
Dear Cagri,

I've been struggling with an error when running the program on the normal files (RNN_Train.py). 
I want to get the F1-score and implemented it the same way as in the "RNN_First_Test.py" file. I always get this error at the end. 

_ValueError: Mix of label input types (string and number)_

Is there any solution to this error? I tried converting the elements inside the predictors but that didn't work. 

Thanks for the help.


EDIT: Found the error and corrected it. Seems to work.
